### ISSUE TITLE
Caruana, Guos algoritm for Gaussian guess function?

### ISSUE THREAD

**Issue by dtnguyen2** [sherpa/sherpa#875]
While working on [PR#741](https://github.com/sherpa/sherpa/pull/741), I noticed that the underlying 1D ```guess``` algorithm for the Gaussian function is not very sophisticated (the location of the peak value is the position, the peak value is the amplitude and to find the fwhm one marches off from the peak value until half the peak value is found etc...) but it is probably good enough for most cases.  

However, Raphael is making good progress on MLE by using simultaneous fitting of crowded fields.  In his prototype, he no longer uses the ```wavedect``` positions as the initial positions for the fitted parameters since they sometimes prove to be unreliable but he uses the sherpa built-in ```guess``` function.  If ever the Gaussian guess has to be improved for MLE, I suggest we look into the _Caruana_ and _Guos_ algorithm.

Moreover, the current sherpa 2D Gaussian guess function does not calculate the two ```sigmas``` so one could in theory "sum up" the bins in the x and y axes to reduce the 2D Gaussian to two 1D Gaussians and apply the aforementioned algorithm.

**Comment by DougBurke** [comment]
@dtnguyen2 - is https://content.sciendo.com/view/journals/jee/66/3/article-p178.xml a good reference? Fortunately it's open access.

**Comment by dtnguyen2** [comment]
Yes, the reference [Caruana, Guos algorithm](https://content.sciendo.com/view/journals/jee/66/3/article-p178.xml) is a clear enough to implement from

**Comment by DougBurke** [comment]
Just to note that I've been tracking down a difference between CIAO 4.12 and 4.13 and one of them was the change in behavior of the guess algorithm for the gauss1d.fwhm component. From https://github.com/sherpa/sherpa/commit/9c29df316488d9f3114ec3adc064d4c863c19baa (#741) the changes to `sherpa/utils/__init__.py.get_fwhm`

a) the code now goes to the right of the peak whereas it started from the left-edge before [there's no a-priori right way, and this version is at least better-connected to the peak whereas the old version would just stop at the first value found, even if not associated to the peak]

We could perhaps go from both sides and take min, max, or mean of the two widths.

b) the old version returned

            return 2.0 * numpy.abs(x[ii] - x_max)	

whereas the new version returns

            return 2.0 * ii

I think it should be returning the old version (so `2 * abs(x[ii] -xmax`), as `ii` is just an index. That is, instead of

    for ii, val in enumerate(y[y_argmax:]):
        if val < half_max_val:
            return 2.0 * ii

we should have (untested)

    for ii, val in enumerate(y[y_argmax:]):
        if val < half_max_val:
            return 2.0 * (x[ii] - x_max)
 
Although this should be replaced by the Caruna/Guos algorithm (if it performs better).

**Comment by DougBurke** [comment]
Actually, should be mode like `(x[ii + y_argmax] - x_max)` or, better still, just send in the x value too.

**Comment by DougBurke** [comment]
Also, the 2D guess - in this case `sherpa.models.basic.Gauss2D` has

```
    def guess(self, dep, *args, **kwargs):
        xpos, ypos = guess_position(dep, *args)
        norm = guess_amplitude2d(dep, *args)
        fwhm = guess_fwhm(dep, *args)
        param_apply_limits(xpos, self.xpos, **kwargs)
        param_apply_limits(ypos, self.ypos, **kwargs)
        param_apply_limits(norm, self.ampl, **kwargs)
        param_apply_limits(fwhm, self.fwhm, **kwargs)
```

where this calls `sherpa.utils.guess_fwhm`, **BUT** this function assumes a 1D dataset, so the Y values are ignored (it essentially uses a 1D version of the image data, collapsed along the Y axis).

**Comment by DougBurke** [comment]
So #1031 improves the FWHM algorithm to address the 1D case. It does nothing about the 2D guess issue. I'd consider this a back-up for the Caruana/Guos algorithm (but has the advantage of having a PR).

**Comment by dtnguyen2** [comment]
For the 2D guess, how about summing up along the x and y axes to get two simulated 1D cases then use the FWHM algorithm from the 1D case?
